pulsar-common/src/main/java/org/apache/pulsar/common/protocol/ByteBufPair.java

```diff
@@ -122,38 +123,22 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
                 // ByteBuf are automatically released after a write. If the ByteBufPair ref count is increased and it
                 // gets written multiple times, the individual buffers refcount should be reflected as well.
                 try {
-                    ctx.write(b.getFirst().retainedDuplicate(), ctx.voidPromise());
-                    ctx.write(b.getSecond().retainedDuplicate(), promise);
+                    ctx.write(readOnlyRetainedDuplicate(b.getFirst()), ctx.voidPromise());
+                    ctx.write(readOnlyRetainedDuplicate(b.getSecond()), promise);
                 } finally {
                     ReferenceCountUtil.safeRelease(b);
                 }
             } else {
                 ctx.write(msg, promise);
             }
         }
-    }
-
-    @Sharable
-    @SuppressWarnings("checkstyle:JavadocType")
-    public static class CopyingEncoder extends ChannelOutboundHandlerAdapter {
-        @Override
-        public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
-            if (msg instanceof ByteBufPair) {
-                ByteBufPair b = (ByteBufPair) msg;
 
-                // Some handlers in the pipeline will modify the bytebufs passed in to them (i.e. SslHandler).
-                // For these handlers, we need to pass a copy of the buffers as the source buffers may be cached
-                // for multiple requests.
-                try {
-                    ctx.write(b.getFirst().copy(), ctx.voidPromise());
-                    ctx.write(b.getSecond().copy(), promise);
-                } finally {
-                    ReferenceCountUtil.safeRelease(b);
-                }
-            } else {
-                ctx.write(msg, promise);
-            }
+        // .asReadOnly() is needed to prevent SslHandler from modifying the input buffers.
+        private static ByteBuf readOnlyRetainedDuplicate(ByteBuf buf) {
+            // If the buffer is already read-only, .asReadOnly() will return the same buffer.
+            // That's why the additional .retainedDuplicate() is needed to ensure that the returned buffer
+            // has independent readIndex and writeIndex.
+            return buf.asReadOnly().retainedDuplicate();
```

Comment:
```java
    private static ByteBuf readOnlyRetainedDuplicate(ByteBuf buf) {
        if (buf == null || buf.readableBytes() <= 0) {
            return Unpooled.EMPTY_BUFFER;
        }
        
        if (buf.isReadOnly()) {
            return buf.retainedDuplicate().asReadOnly();
        }
        return buf.asReadOnly().retain();
    }
```

Is this should be better? I think we don't need to call `duplicate()` after call `asReadOnly`

Reply:
> I think we don't need to call `duplicate()` after call `asReadOnly`

It is needed. The comment explains it. please check the comment for the reason.

Reply:
I mean if the input buf is not `readOnly`, `asReadOnly` will create a new instance, so we just need to `retain()` after `asReadOnly`.
It also has independent read/write index.

Reply:
> if (buf == null || buf.readableBytes() <= 0) {
>            return Unpooled.EMPTY_BUFFER;
>        }

I don't think that nulls should be tolerated when nulls shouldn't be passed as input. For the readableBytes() == 0 optimization, is that needed?

Reply:
This is just defensive programming, if we are sure that the input buf cannot be null, this line can be removed.

Reply:
This approach is currently blocked by another Netty bug where SslHandler doesn't support a read only buffer. That seems to be the reason why a deep copy is currently required. Fixing that issue in netty/netty#14071.

Reply:
Oh, I checked the source code, it looks a little strange that I didn't understand at the first time.

```java
    private static boolean attemptCopyToCumulation(ByteBuf cumulation, ByteBuf next, int wrapDataSize) {
        final int inReadableBytes = next.readableBytes();
        final int cumulationCapacity = cumulation.capacity();
        if (wrapDataSize - cumulation.readableBytes() >= inReadableBytes &&
                // Avoid using the same buffer if next's data would make cumulation exceed the wrapDataSize.
                // Only copy if there is enough space available and the capacity is large enough, and attempt to
                // resize if the capacity is small.
                (cumulation.isWritable(inReadableBytes) && cumulationCapacity >= wrapDataSize ||
                        cumulationCapacity < wrapDataSize &&
                                ensureWritableSuccess(cumulation.ensureWritable(inReadableBytes, false)))) {
            cumulation.writeBytes(next);
            next.release();
            return true;
        }
        return false;
    }
```
We should return false immediately if `cumulation` is not writable

Reply:
> This approach is currently blocked by another Netty bug where SslHandler doesn't support a read only buffer. That seems to be the reason why a deep copy is currently required. Fixing that issue in [netty/netty#14071](https://github.com/netty/netty/pull/14071).

Yes, the fix looks good

Reply:
> We should return false immediately if cumulation is not writable

In this case, isWritable returns true for a buffer for a wrapped read only buffer. This is surprising and the reason why this read only buffers aren't supported if there's another wrapper. It correctly returns true for isReadOnly.

One possible workaround could be to use `Unpooled.unmodifiableBuffer` to add the readonly wrapper so that it's always the "top most" wrapper. That would also avoid the need for the extra `duplicate()` wrapper.

Reply:
> In this case, isWritable returns true for a buffer for a wrapped read only buffer.
It sounds very strange, can you please point me where?

Reply:
> It sounds very strange, can you please point me where?

Yes, it's very surprising behavior. You can try it out with a debugger by running for example TlsProducerConsumerTest and checking what happens in the attemptCopyToCumulation if the `.asReadOnly().retainedDuplicate()` solution is used. 

The io.netty.buffer.AbstractByteBuf#isWritable(int) method gets called: https://github.com/netty/netty/blob/70d6a3f40d7e8fd3f5ced7600ed209c58944f673/buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java#L171-L174
It will evaluate to true since it's implementation doesn't check for `isReadOnly()`:
```java
    @Override
    public boolean isWritable(int numBytes) {
        return capacity() - writerIndex >= numBytes;
    }
```